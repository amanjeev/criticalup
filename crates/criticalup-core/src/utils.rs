// SPDX-FileCopyrightText: The Ferrocene Developers
// SPDX-License-Identifier: MIT OR Apache-2.0

use crate::errors::WriteFileError;
use sha2::{Digest, Sha256};
use std::hash::Hasher;
use std::path::Path;
use tokio::fs::File;
use tokio::io::BufWriter;

pub(crate) async fn open_file_for_write(path: &Path) -> Result<BufWriter<File>, WriteFileError> {
    // Ensure the parent directory is always present
    if let Some(parent) = path.parent() {
        tokio::fs::create_dir_all(parent)
            .await
            .map_err(WriteFileError::CantCreateParentDirectory)?;
    }

    Ok(BufWriter::new(
        File::create(path).await.map_err(WriteFileError::Io)?,
    ))
}

/// A `Hasher` helper type which is a wrapper to a choice of cryptographic hashing algorithm
/// to generate cryptographic hash of our types. This is needed to make sure we
/// 1. do not use [`DefaultHasher`], which may change its algorithm, for hash state
/// 2. bridge the gap between normal [`Hash`] and cryptographic hash (e.g. [`Sha256`])
/// 3. better ergonomics to create a hash of our types like [`ProjectManifestProduct`] using `#[derive(Hash)]`
pub struct Sha256Hasher(Sha256);

impl Sha256Hasher {
    pub(crate) fn new() -> Self {
        Self(Sha256::new())
    }

    /// Provides the final hash value
    pub(crate) fn finalize(self) -> String {
        format!("{:x}", self.0.finalize())
    }
}

impl Hasher for Sha256Hasher {
    /// This method is unreachable and here to appease the compiler, mandatory method.
    fn finish(&self) -> u64 {
        unreachable!()
    }

    /// Update the hasher state, mandatory method.
    fn write(&mut self, bytes: &[u8]) {
        self.0.update(bytes)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::hash::Hash;

    #[test]
    fn test_sha256_for_a_struct_works() {
        #[derive(Hash)]
        struct Abc {
            name: String,
            version: String,
            dependencies: Vec<String>,
        }

        let abc = Abc {
            name: "abc".to_string(),
            version: "1.2.3".to_string(),
            dependencies: vec!["dep2".to_string(), "dep1".to_string()],
        };

        let mut hasher = Sha256Hasher::new();
        abc.hash(&mut hasher);
        let final_hash = hasher.finalize();
        assert_eq!(
            "fb9eee112b5cee551f7a5088402e53dbbdaec2a1e1cd3f4663a1b81b1b53015f",
            final_hash
        )
    }
}
