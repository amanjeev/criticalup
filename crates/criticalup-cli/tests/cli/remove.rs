use crate::assert_output;
use crate::utils::{construct_toolchains_product_path, TestEnvironment};
use serde_json::{json, Value};
use std::fs::File;
use std::io::{BufReader, Write};

#[test]
fn help_message() {
    let test_env = TestEnvironment::prepare();
    assert_output!(test_env.cmd().args(["remove", "--help"]));
}

#[test]
fn remove_deletes_only_manifest_from_list_and_dir() {
    let test_env = TestEnvironment::prepare();
    let mut current_dir = std::env::current_dir().unwrap();
    current_dir.push("tests/resources/criticalup.toml");
    let manifest_path = current_dir.canonicalize().unwrap();

    // Generate the manifest object so we can get the installation id hash.
    let manifest =
        criticalup_core::project_manifest::ProjectManifest::load(current_dir.as_path()).unwrap();
    let installation_id = manifest.products()[0].installation_id();

    let state_file_path = test_env.root().join("state.json");
    let mut state_file = File::create(&state_file_path).unwrap();
    // 6bb4fe4c8205d18a8eaf0b852c3b29f65805fd80e528af74cf2f1463a911e40e is the hash of the
    // current criticalup.toml's product contents which we use here to create state.json by
    // dynamically calculating it from the criticalup.toml.
    let content = json!(
        { "version": 1,
          "authentication_token": "criticalup_token_45_hahaha",
          "installations": {
            &installation_id.0: {
              "binary_proxies": {
                "cargo": "/path/toolchains/bin/cargo"
               },
              "manifests": [
                    manifest_path
              ]
            }
          }
        }
    )
    .to_string();

    state_file.write_all(content.as_bytes()).unwrap();
    // Manually create the toolchain directory which allows us to skip installation.
    let product_toolchain_dir =
        construct_toolchains_product_path(&test_env, installation_id.0.as_str());
    std::fs::create_dir_all(&product_toolchain_dir).unwrap();

    assert_output!(test_env
        .cmd()
        .args(["remove", "--project", manifest_path.to_str().unwrap()]));

    let state_file_actual: Value =
        serde_json::from_reader(BufReader::new(File::open(&state_file_path).unwrap())).unwrap();

    // Installation's manifest is an empty array because the manifest path was removed.
    assert_eq!(
        state_file_actual
            .pointer(format!("/installations/{}/manifests", installation_id.0).as_str())
            .unwrap(),
        &json!([])
    );

    // Directory is gone.
    assert!(!product_toolchain_dir.exists());
}
