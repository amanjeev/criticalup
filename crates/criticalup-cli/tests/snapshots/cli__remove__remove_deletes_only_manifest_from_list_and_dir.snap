---
source: crates/criticalup-cli/tests/cli/remove.rs
expression: repr
---
exit: exit status: 0

stdout
------
[1minfo:[0m deleting installation 6bb4fe4c8205d18a8eaf0b852c3b29f65805fd80e528af74cf2f1463a911e40e
------

empty stderr
